token-lending/program/src/state/obligation.rs

```diff
@@ -1,186 +1,189 @@
+    ) -> Result<(&ObligationCollateral, usize), ProgramError> {
+        if self.collateral.is_empty() {
+            return Err(LendingError::ObligationCollateralEmpty.into());
+        }
+        let index = self
+            ._find_collateral_index(deposit_reserve)
+            .ok_or(LendingError::InvalidObligationCollateral)?;
+        Ok((&self.collateral[index], index))
     }
 
-    /// Accrue interest
-    pub fn accrue_interest(&mut self, cumulative_borrow_rate: Decimal) -> ProgramResult {
-        if cumulative_borrow_rate < self.cumulative_borrow_rate_wads {
-            return Err(LendingError::NegativeInterestRate.into());
+    /// Find or add collateral by deposit reserve
+    pub fn find_or_add_collateral(
+        &mut self,
+        deposit_reserve: Pubkey,
+    ) -> Result<&mut ObligationCollateral, ProgramError> {
+        if let Some(index) = self._find_collateral_index(deposit_reserve) {
+            Ok(&mut self.collateral[index])
+        } else if self.collateral.len() + self.liquidity.len() >= MAX_OBLIGATION_RESERVES {
+            Err(LendingError::ObligationReserveLimit.into())
+        } else {
+            let collateral = ObligationCollateral::new(deposit_reserve);
+            self.collateral.push(collateral);
+            Ok(self.collateral.last_mut().unwrap())
         }
+    }
 
-        let compounded_interest_rate: Rate = cumulative_borrow_rate
-            .try_div(self.cumulative_borrow_rate_wads)?
-            .try_into()?;
-
-        self.borrowed_liquidity_wads = self
-            .borrowed_liquidity_wads
-            .try_mul(compounded_interest_rate)?;
-
-        self.cumulative_borrow_rate_wads = cumulative_borrow_rate;
-
-        Ok(())
+    fn _find_collateral_index(&self, deposit_reserve: Pubkey) -> Option<usize> {
+        self.collateral
+            .iter()
+            .position(|collateral| collateral.deposit_reserve == deposit_reserve)
     }
 
-    /// Liquidate part of obligation
-    pub fn liquidate(&mut self, repay_amount: Decimal, withdraw_amount: u64) -> ProgramResult {
-        self.borrowed_liquidity_wads = self.borrowed_liquidity_wads.try_sub(repay_amount)?;
-        self.deposited_collateral_tokens = self
-            .deposited_collateral_tokens
-            .checked_sub(withdraw_amount)
-            .ok_or(LendingError::MathOverflow)?;
-        Ok(())
+    /// Find liquidity by borrow reserve
+    pub fn find_liquidity(
+        &self,
+        borrow_reserve: Pubkey,
+    ) -> Result<(&ObligationLiquidity, usize), ProgramError> {
+        if self.liquidity.is_empty() {
+            return Err(LendingError::ObligationLiquidityEmpty.into());
+        }
+        let index = self
+            ._find_liquidity_index(borrow_reserve)
+            .ok_or(LendingError::InvalidObligationLiquidity)?;
+        Ok((&self.liquidity[index], index))
     }
 
-    /// Repay borrowed tokens
-    pub fn repay(
+    /// Find or add liquidity by borrow reserve
+    pub fn find_or_add_liquidity(
```

Comment:
Yeah, I think collateral withdraw, repayment, or liquidation that zeros out the amount should automatically remove it, or it could be done on refresh.

Reply:
Yeah, having an explicit instruction wouldn't be bad either. I was just thinking ahead that it's not super great that borrow and deposit state are different sizes. I wish they were all aligned so that this zeroing out behavior wouldn't cause the obligation state storage to get fragmented

Reply:
Hmm, can you explain what you mean by fragmented?

Reply:
This is done, implemented by immediately removing the collateral or liquidity when zeroed out.

Obligations are packed with all deposits in order followed by all borrows in order, and then there's just whatever empty space left over at the end. There always has to be at least one collateral before there can be any borrows. If a repayment wipes the last borrow, there will just be one deposit which can then be withdrawn and wiped. If a liquidation wipes the last borrow and deposit together, the obligation will be in the same state as when it was initialized.

I don't see what difference it would make if they were the same size though.

Reply:
Ah never mind, I didn't realize that all of the deposits and borrows are rewritten to account data each time. Your approach is fine for now, if we need to optimize writes in the future we can revisit